### DISCUSSION TITLE
Support for syntax formatting in executable listings?

### DISCUSSION THREAD

**Discussion by pat-alt** [quarto-dev/quarto-cli#7610] [Q&A]
### Description

The current [prerelease](https://quarto.org/docs/prerelease/1.4/crossref.html#cross-referenceable-listings-of-executable-code-blocks) offers support for cross-referenceable listings of executable code blocks. Is it or will it be possible to format these?

*Specific Example*

I'm working on a Quarto extension for [JuliaCon Proceedings](https://proceedings.juliacon.org/) here: [https://github.com/pat-alt/quarto-juliacon-proceedings](https://github.com/pat-alt/quarto-juliacon-proceedings). The existing [LaTeX template](https://github.com/JuliaCon/JuliaConSubmission.jl) ships with a custom [style file](https://github.com/JuliaCon/JuliaConSubmission.jl/blob/master/paper/jlcode.sty) that defines a special environment for Julia code, built on top of `listings`. Would it be possible to somehow specify that I want to use that environment?

The [template](https://github.com/pat-alt/quarto-juliacon-proceedings/blob/main/template.qmd) contains an [example](https://github.com/pat-alt/quarto-juliacon-proceedings/blob/7284d862f246a5749d1fe68880a27487aa233a99/template.qmd#L84) using a listing of executable code and an [example](https://github.com/pat-alt/quarto-juliacon-proceedings/blob/7284d862f246a5749d1fe68880a27487aa233a99/template.qmd#L109) using LaTeX with the custom environment. The rendered output can be found here: [pdf](https://www.paltmeyer.com/quarto-juliacon-proceedings/template.pdf), [html](https://www.paltmeyer.com/quarto-juliacon-proceedings/template.html).

**Comment by cscheid** [comment]
Yes, it should be possible for your extension to detect those listings in a filter, and then annotate the AST so that they're rendered how you want them to. I'd be happy to help you with it. Feel free to ask further questions here on the forums or send me an email directly at carlos.scheidegger@posit.co

**Reply by pat-alt** [reply]
Thanks very much @cscheid! 

I've managed to get pretty close but would indeed much appreciate your help with some remaining issues.

The filter I've added [here](https://github.com/pat-alt/quarto-juliacon-proceedings/pull/27/files#diff-2be545b2649c5b39fee8171138ded7a97ee71e828013945f535f415f9893d980) works as expected for executable Julia code chunks. But when I add `lst-label` and `lst-caption` options, the filter takes no effect. I think this is because the functionality shipped with the [prerelease](https://quarto.org/docs/prerelease/1.4/crossref.html#cross-referenceable-listings-of-executable-code-blocks) uses its own custom `codelisting` environment? Below are the code chunks and their corresponding raw latex output for both options. The rendered PDF is [here](https://github.com/pat-alt/quarto-juliacon-proceedings/blob/9-better-syntax-highlighting/docs/template.pdf).

#### Option 1

No `lst-label` and `lst-caption` and hence not cross-referenceable:

```{julia}
#| label: fig-2
#| fig-cap: A figure caption.
#| eval: true
#| echo: true

using Plots

x = -3.0:0.01:3.0
y = cumsum(rand(length(x)))
plot(x, y)
```

Yields:

```tex
\begin{lstlisting}[language=Julia]
using Plots

x = -3.0:0.01:3.0
y = cumsum(rand(length(x)))
plot(x, y)
\end{lstlisting}
```

#### Option 2

Cross-referenceable:

```{julia}
#| lst-label: lst-1
#| lst-cap: A listing caption.
#| label: fig-1
#| fig-cap: A figure caption.
#| eval: true
#| echo: true

using Plots

x = -3.0:0.01:3.0
y = rand(length(x))
plot(x, y)
```

Yields:

```tex
\begin{codelisting}

\caption{\label{lst-1}A listing caption.}

\centering{

\begin{verbatim}
using Plots

x = -3.0:0.01:3.0
y = rand(length(x))
plot(x, y)
\end{verbatim}

}

\end{codelisting}
```

- Can I somehow make this work with the `codelisting` environment?
- Any idea how I might still be able to pass on additional options for `lstlisting` like `numbers=left`? I can do this by default in the lua filter but ideally I want to be able to provide this through Quarto/yml.

**Reply by pat-alt** [reply]
@cscheid friendly bump :) maybe easier if I indeed drop you an email? Would be amazing to get this sorted in the new year. For now, happy holidays!

**Reply by cscheid** [reply]
> maybe easier if I indeed drop you an email?

Yes, please, and thank you.
